src/meltano/core/manifest/manifest.py

```diff
@@ -0,0 +1,422 @@
+
+    Args:
+        data: The dictionary being merged into.
+        key: The key for the dictionary being merged into.
+        value: The value being merged into the dictionary.
+        _: Merge strategies - unused argument.
+
+    Returns:
+        `NotImplemented` if the key is not "env"; `None` otherwise.
+    """
+    if key != "env":
+        return NotImplemented  # type: ignore
+    data[key] = {
+        **expand_env_vars(data[key], value, if_missing=EnvVarMissingBehavior.ignore),
+        **expand_env_vars(value, data[key], if_missing=EnvVarMissingBehavior.ignore),
+    }
+
+
+class Manifest:
+    """A complete unambiguous static representation of a Meltano project."""
+
+    def __init__(
+        self,
+        project: Project,
+        environment: Environment | None,
+        path: Path,
+    ) -> None:
+        """Initialize the manifest.
+
+        This class does not write a manifest file. It merely generates the data
+        that would go into one.
+
+        The contents of the manifest can be access using the `data` attribute.
+        This data is generated when it is first accessed. After that, a cached
+        result is used.
+
+        Args:
+            project: The Meltano project which this manifest describes.
+            environment: The Meltano environment which this manifest describes.
+                If `None`, then all data under the `environments` key in the
+                project files will be ignored by this manifest.
+            path: The path the Manifest will be saved to - only used for
+                jsonschema validation failure error messages.
+        """
+        self.project = project
+        self._meltano_file = self.project.meltanofile.read_text()
+        self.environment = environment
+        self.path = path
+        self.project_settings_service = ProjectSettingsService(project)
+        self.project_plugins_service = ProjectPluginsService(project)
+        with open(MANIFEST_SCHEMA_PATH) as manifest_schema_file:
+            self.manifest_schema = json.load(manifest_schema_file)
+
+    @staticmethod
+    def _validate_against_manifest_schema(
+        instance_name: str,
+        instance_path: Path,
+        intance_data: dict[str, Any],
+    ) -> None:
+        with NamedTemporaryFile(suffix=".json") as schema_instance_file:
+            schema_instance_file.write(json.dumps(intance_data).encode())
+            schema_instance_file.flush()
+            proc = subprocess.run(
+                (
+                    "check-jsonschema",
+                    "--color",
+                    "never" if get_no_color_flag() else "always",
+                    "--schemafile",
+                    str(MANIFEST_SCHEMA_PATH),
+                    schema_instance_file.name,
+                ),
+                text=True,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+            )
```

Comment:
I don't recall a discussion of `check-jsonschema` vs the `jsonschema` Python API 😳 
 . I guess there's a (probably small) performance hit from invoking a subprocess here so are there significant benefits over using [`iter_errors`](https://python-jsonschema.readthedocs.io/en/stable/api/jsonschema/#jsonschema.Validator.iter_errors) other than getting an already formatted output from that CLI?

Reply:
I don't think we ever discussed this. This is an example of what it looks like when using `iter_errors`:

Code:

```python
        validator = jsonschema.Draft202012Validator(
            schema=json.loads(MANIFEST_SCHEMA_PATH.read_text())
        )
        errors = list(validator.iter_errors(instance=instance_data))
        if errors:
            logger.warn(
                f"Failed to validate {instance_name} against Meltano manifest "
                f"schema ({MANIFEST_SCHEMA_PATH}):\n"
                + "\n".join(str(error) for error in errors)
            )
```

Output:

```
2023-02-03T23:33:08.409688Z [warning  ] Failed to validate newly compiled manifest against Meltano manifest schema (/home/will/meltano/meltano2/src/meltano/schemas/meltano.schema.json):
Unevaluated properties are not allowed ('description', 'plugin_type', 'repo' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['extractors']['items']:
    {'allOf': [{'$ref': '#/$defs/plugins/plugin_generic'},
               {'$ref': '#/$defs/plugins/extractor_shared'},
               {'properties': {'capabilities': {'items': {'enum': ['properties',
                                                                   'catalog',
                                                                   'discover',
                                                                   'state',
                                                                   'about',
                                                                   'stream-maps',
                                                                   'activate-version',
                                                                   'batch',
                                                                   'test',
                                                                   'log-based'],
                                                          'type': 'string'},
                                                'type': 'array'}},
                'type': 'object'}],
     'description': 'https://docs.meltano.com/concepts/plugins#extractors',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['extractors'][0]:
    {'capabilities': ['catalog', 'discover', 'state'],
     'description': 'Single application for the entire DevOps lifecycle',
     'docs': 'https://hub.meltano.com/extractors/tap-gitlab--meltanolabs',
     'env': {'EB': 'EB overwritten',
             'EC': '${EB}CCC',
             'MELTANO_EXTRACT_API_URL': 'https://gitlab.com',
             'MELTANO_EXTRACT_FETCH_MERGE_REQUEST_COMMITS': 'false',
             'MELTANO_EXTRACT_FETCH_PIPELINES_EXTENDED': 'false',
             'MELTANO_EXTRACT_GROUPS': '',
             'MELTANO_EXTRACT_PRIVATE_TOKEN': '',
             'MELTANO_EXTRACT_PROJECTS': '',
             'MELTANO_EXTRACT_ULTIMATE_LICENSE': 'false',
             'MELTANO_EXTRACT__LOAD_SCHEMA': 'tap_gitlab',
             'MELTANO_EXTRACT__METADATA': '{}',
             'MELTANO_EXTRACT__SCHEMA': '{}',
             'MELTANO_EXTRACT__SELECT': '["*.*"]',
             'MELTANO_EXTRACT__SELECT_FILTER': '[]',
             'TAP_GITLAB_API_URL': 'https://gitlab.com',
             'TAP_GITLAB_FETCH_MERGE_REQUEST_COMMITS': 'false',
             'TAP_GITLAB_FETCH_PIPELINES_EXTENDED': 'false',
             'TAP_GITLAB_GROUPS': '',
             'TAP_GITLAB_PRIVATE_TOKEN': '',
             'TAP_GITLAB_PROJECTS': '',
             'TAP_GITLAB_ULTIMATE_LICENSE': 'false',
             'TAP_GITLAB__LOAD_SCHEMA': 'tap_gitlab',
             'TAP_GITLAB__METADATA': '{}',
             'TAP_GITLAB__SCHEMA': '{}',
             'TAP_GITLAB__SELECT': '["*.*"]',
             'TAP_GITLAB__SELECT_FILTER': '[]'},
     'label': 'GitLab',
     'logo_url': 'https://hub.meltano.com/assets/logos/extractors/gitlab.png',
     'name': 'tap-gitlab',
     'namespace': 'tap_gitlab',
     'pip_url': 'git+https://github.com/MeltanoLabs/tap-gitlab.git',
     'plugin_type': 'extractors',
     'repo': 'https://github.com/MeltanoLabs/tap-gitlab',
     'settings': [{'description': 'GitLab API/instance URL. When an API '
                                  'path is omitted, `/api/v4/` is assumed.',
                   'label': 'GitLab Instance',
                   'name': 'api_url',
                   'protected': True,
                   'value': 'https://gitlab.com'},
                  {'description': 'GitLab personal access token or other '
                                  'API token.\n'
                                  '\n'
                                  '#### How to get\n'
                                  '\n'
                                  'The process for getting the private '
                                  'token or personal access token is very '
                                  'simple:\n'
                                  '\n'
                                  '<video controls style="max-width: '
                                  '100%">\n'
                                  '  <source '
                                  'src="/assets/videos/tap-gitlab/personal-access-token.mov">\n'
                                  '</video>\n'
                                  '\n'
                                  "1. Navigate to your [profile's access "
                                  'tokens](https://gitlab.com/-/profile/personal_access_tokens).\n'
                                  '\n'
                                  '2. Fill out the personal access token '
                                  'form with the following properties:\n'
                                  '\n'
                                  '- **Name:** meltano-gitlab-tutorial\n'
                                  '- **Expires:** _leave blank unless you '
                                  'have a specific reason to expire the '
                                  'token_\n'
                                  '- **Scopes:**\n'
                                  '  - api\n'
                                  '\n'
                                  '3. Click on `Create personal access '
                                  'token` to submit your request.\n'
                                  '\n'
                                  '4. You should see your token appear at '
                                  'the top of your screen. It should look '
                                  'something like this: '
                                  '`I8vxHsiVAaDnAX3hA`\n',
                   'kind': 'password',
                   'label': 'Access Token',
                   'name': 'private_token',
                   'placeholder': 'Ex. *****************',
                   'value': ''},
                  {'description': 'This property allows you to scope data '
                                  'that the extractor fetches to only the '
                                  'desired group. The group name can '
                                  'generally be found at the root of a '
                                  "repository's URL. If this is left "
                                  'blank, you have to at least provide a '
                                  '[project](#projects).\n'
                                  '\n'
                                  "Leave empty if you'd like to pull data "
                                  'from a project in a personal user '
                                  'namespace.\n'
                                  '\n'
                                  'For example, '
                                  '`https://github.com/MeltanoLabs/tap-gitlab` '
                                  'has a group of `meltano`.\n'
                                  '\n'
                                  'Multiple group names can be separated '
                                  'using space characters.\n',
                   'label': 'Groups',
                   'name': 'groups',
                   'placeholder': 'Ex. my-organization',
                   'value': ''},
                  {'description': 'This property allows you to scope the '
                                  'project(s) that the extractor fetches.\n'
                                  '\n'
                                  "Leave empty if you've specified one or "
                                  'more [groups](#groups) and would like '
                                  'to pull data from all projects inside '
                                  'these groups.\n'
                                  '\n'
                                  'The format for it is '
                                  '`namespace/project`, where namespace '
                                  'can be a username or group name. Here '
                                  'are a couple examples:\n'
                                  '\n'
                                  '- `meltano/meltano` - The core [Meltano '
                                  'project](https://gitlab.com/meltano/meltano)\n'
                                  '- `meltano/sdk` - The project for the '
                                  '[Meltano SDK for Singer Taps and '
                                  'Targets '
                                  '](https://gitlab.com/meltano/sdk)\n'
                                  '\n'
                                  'Multiple group paths can be separated '
                                  'using space characters.\n',
                   'label': 'Project',
                   'name': 'projects',
                   'placeholder': 'Ex. my-organization/project-1',
                   'value': ''},
                  {'description': 'Enable to pull in extra data (like '
                                  'Epics, Epic Issues and other entities)\n'
                                  'only available to GitLab Ultimate and '
                                  'GitLab.com Gold accounts.\n'
                                  '\n'
                                  'The `epics` and `epic_issues` entities '
                                  'cannot be '
                                  '[selected](https://docs.meltano.com/guide/integration#selecting-entities-and-attributes-for-extraction) '
                                  'unless this setting is enabled.\n',
                   'kind': 'boolean',
                   'label': 'Ultimate License',
                   'name': 'ultimate_license',
                   'value': False},
                  {'description': 'For each Merge Request, also fetch the '
                                  "MR's commits and create the join table "
                                  '`merge_request_commits` with the Merge '
                                  'Request and related Commit IDs.\n'
                                  '\n'
                                  'This can slow down extraction '
                                  'considerably because of the many API '
                                  'calls required.\n'
                                  '\n'
                                  'The `merge_request_commits` entity '
                                  'cannot be '
                                  '[selected](https://docs.meltano.com/guide/integration#selecting-entities-and-attributes-for-extraction) '
                                  'unless this setting is enabled.\n',
                   'kind': 'boolean',
                   'label': 'Fetch Merge Request Commits',
                   'name': 'fetch_merge_request_commits',
                   'value': False},
                  {'description': 'For every Pipeline, also fetch extended '
                                  'details of each of these pipelines.\n'
                                  '\n'
                                  'This  can slow down extraction '
                                  'considerably because of the many API '
                                  'calls required.\n'
                                  '\n'
                                  'The `pipelines_extended` entity cannot '
                                  'be '
                                  '[selected](https://docs.meltano.com/guide/integration#selecting-entities-and-attributes-for-extraction) '
                                  'unless this setting is enabled.\n',
                   'kind': 'boolean',
                   'label': 'Fetch Pipelines Extended',
                   'name': 'fetch_pipelines_extended',
                   'value': False},
                  {'description': 'Determines how much historical data '
                                  'will be extracted. Please be aware that '
                                  'the larger the time period and amount '
                                  'of data, the longer the initial '
                                  'extraction can be expected to take.',
                   'kind': 'date_iso8601',
                   'label': 'Start Date',
                   'name': 'start_date'}],
     'settings_group_validation': [['api_url', 'groups', 'start_date'],
                                   ['api_url', 'projects', 'start_date']],
     'variant': 'meltanolabs'}
Unevaluated properties are not allowed ('plugin_type', 'repo' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['extractors']['items']:
    {'allOf': [{'$ref': '#/$defs/plugins/plugin_generic'},
               {'$ref': '#/$defs/plugins/extractor_shared'},
               {'properties': {'capabilities': {'items': {'enum': ['properties',
                                                                   'catalog',
                                                                   'discover',
                                                                   'state',
                                                                   'about',
                                                                   'stream-maps',
                                                                   'activate-version',
                                                                   'batch',
                                                                   'test',
                                                                   'log-based'],
                                                          'type': 'string'},
                                                'type': 'array'}},
                'type': 'object'}],
     'description': 'https://docs.meltano.com/concepts/plugins#extractors',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['extractors'][1]:
    {'capabilities': ['properties', 'discover', 'state'],
     'docs': 'https://hub.meltano.com/extractors/tap-adwords--singer-io',
     'env': {'MELTANO_EXTRACT_CONVERSION_WINDOW_DAYS': '0',
             'MELTANO_EXTRACT_PRIMARY_KEYS': '{"KEYWORDS_PERFORMANCE_REPORT": '
                                             '["customerID", "campaignID", '
                                             '"adGroupID", "keywordID", '
                                             '"day", "network", "device"], '
                                             '"AD_PERFORMANCE_REPORT": '
                                             '["customerID", "campaignID", '
                                             '"adGroupID", "adID", "day", '
                                             '"network", "device"]}',
             'MELTANO_EXTRACT_USER_AGENT': 'tap-adwords via Meltano',
             'MELTANO_EXTRACT__LOAD_SCHEMA': 'tap_adwords',
             'MELTANO_EXTRACT__METADATA': '{}',
             'MELTANO_EXTRACT__SCHEMA': '{}',
             'MELTANO_EXTRACT__SELECT': '["campaigns.*", "ad_groups.*", '
                                        '"ads.*", "accounts.*", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.customerID", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.account", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.currency", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.timeZone", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.clientName", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.campaign", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.campaignID", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.campaignState", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.adGroup", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.adGroupID", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.adGroupState", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.day", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.network", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.device", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.clicks", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.cost", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.impressions", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.interactions", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.engagements", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.conversions", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.allConv", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.views", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.activeViewViewableImpressions", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.activeViewMeasurableImpr", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.activeViewMeasurableCost", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.gmailClicksToWebsite", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.gmailSaves", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.gmailForwards", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.keywordID", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.keyword", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.keywordState", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.criterionServingStatus", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.destinationURL", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.matchType", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.topOfPageCPC", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.firstPageCPC", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.imprAbsTop", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.activeViewAvgCPM", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.activeViewViewableCTR", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.activeViewMeasurableImprImpr", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.activeViewViewableImprMeasurableImpr", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.allConvRate", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.allConvValue", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.avgCost", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.avgCPC", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.avgCPE", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.avgCPM", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.avgCPV", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.avgPosition", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.convRate", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.totalConvValue", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.costAllConv", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.costConv", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.costConvCurrentModel", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.crossDeviceConv", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.ctr", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.conversionsCurrentModel", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.convValueCurrentModel", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.engagementRate", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.interactionRate", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.interactionTypes", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.imprTop", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.valueAllConv", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.valueConv", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.valueConvCurrentModel", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo100", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo25", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo50", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo75", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.viewRate", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.viewThroughConv", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.searchAbsTopIS", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.searchLostAbsTopISBudget", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.searchLostTopISBudget", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.searchExactMatchIS", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.searchImprShare", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.searchLostAbsTopISRank", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.searchLostISRank", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.searchLostTopISRank", '
                                        '"KEYWORDS_PERFORMANCE_REPORT.searchTopIS", '
                                        '"AD_PERFORMANCE_REPORT.customerID", '
                                        '"AD_PERFORMANCE_REPORT.account", '
                                        '"AD_PERFORMANCE_REPORT.currency", '
                                        '"AD_PERFORMANCE_REPORT.timeZone", '
                                        '"AD_PERFORMANCE_REPORT.clientName", '
                                        '"AD_PERFORMANCE_REPORT.campaign", '
                                        '"AD_PERFORMANCE_REPORT.campaignID", '
                                        '"AD_PERFORMANCE_REPORT.campaignState", '
                                        '"AD_PERFORMANCE_REPORT.adGroup", '
                                        '"AD_PERFORMANCE_REPORT.adGroupID", '
                                        '"AD_PERFORMANCE_REPORT.adGroupState", '
                                        '"AD_PERFORMANCE_REPORT.day", '
                                        '"AD_PERFORMANCE_REPORT.network", '
                                        '"AD_PERFORMANCE_REPORT.device", '
                                        '"AD_PERFORMANCE_REPORT.clicks", '
                                        '"AD_PERFORMANCE_REPORT.cost", '
                                        '"AD_PERFORMANCE_REPORT.impressions", '
                                        '"AD_PERFORMANCE_REPORT.interactions", '
                                        '"AD_PERFORMANCE_REPORT.engagements", '
                                        '"AD_PERFORMANCE_REPORT.conversions", '
                                        '"AD_PERFORMANCE_REPORT.allConv", '
                                        '"AD_PERFORMANCE_REPORT.views", '
                                        '"AD_PERFORMANCE_REPORT.activeViewViewableImpressions", '
                                        '"AD_PERFORMANCE_REPORT.activeViewMeasurableImpr", '
                                        '"AD_PERFORMANCE_REPORT.activeViewMeasurableCost", '
                                        '"AD_PERFORMANCE_REPORT.gmailClicksToWebsite", '
                                        '"AD_PERFORMANCE_REPORT.gmailSaves", '
                                        '"AD_PERFORMANCE_REPORT.gmailForwards", '
                                        '"AD_PERFORMANCE_REPORT.adID", '
                                        '"AD_PERFORMANCE_REPORT.adState", '
                                        '"AD_PERFORMANCE_REPORT.approvalStatus", '
                                        '"AD_PERFORMANCE_REPORT.adType", '
                                        '"AD_PERFORMANCE_REPORT.adStrength", '
                                        '"AD_PERFORMANCE_REPORT.autoAppliedAdSuggestion", '
                                        '"AD_PERFORMANCE_REPORT.ad", '
                                        '"AD_PERFORMANCE_REPORT.descriptionLine1", '
                                        '"AD_PERFORMANCE_REPORT.descriptionLine2", '
                                        '"AD_PERFORMANCE_REPORT.finalURL", '
                                        '"AD_PERFORMANCE_REPORT.displayURL", '
                                        '"AD_PERFORMANCE_REPORT.description", '
                                        '"AD_PERFORMANCE_REPORT.headline1", '
                                        '"AD_PERFORMANCE_REPORT.headline2", '
                                        '"AD_PERFORMANCE_REPORT.path1", '
                                        '"AD_PERFORMANCE_REPORT.businessName", '
                                        '"AD_PERFORMANCE_REPORT.callToActionTextResponsive", '
                                        '"AD_PERFORMANCE_REPORT.shortHeadline", '
                                        '"AD_PERFORMANCE_REPORT.longHeadline", '
                                        '"AD_PERFORMANCE_REPORT.promotionTextResponsive", '
                                        '"AD_PERFORMANCE_REPORT.responsiveSearchAdPath1", '
                                        '"AD_PERFORMANCE_REPORT.responsiveSearchAdHeadlines", '
                                        '"AD_PERFORMANCE_REPORT.responsiveSearchAdDescriptions", '
                                        '"AD_PERFORMANCE_REPORT.gmailAdBusinessName", '
                                        '"AD_PERFORMANCE_REPORT.gmailAdHeadline", '
                                        '"AD_PERFORMANCE_REPORT.gmailAdDescription", '
                                        '"AD_PERFORMANCE_REPORT.imageAdName", '
                                        '"AD_PERFORMANCE_REPORT.businessNameMultiAssetResponsiveDisplay", '
                                        '"AD_PERFORMANCE_REPORT.longHeadlineMultiAssetResponsiveDisplay", '
                                        '"AD_PERFORMANCE_REPORT.headlinesMultiAssetResponsiveDisplay", '
                                        '"AD_PERFORMANCE_REPORT.callToActionTextMultiAssetResponsiveDisplay", '
                                        '"AD_PERFORMANCE_REPORT.promotionTextMultiAssetResponsiveDisplay", '
                                        '"AD_PERFORMANCE_REPORT.imprAbsTop", '
                                        '"AD_PERFORMANCE_REPORT.activeViewAvgCPM", '
                                        '"AD_PERFORMANCE_REPORT.activeViewViewableCTR", '
                                        '"AD_PERFORMANCE_REPORT.activeViewMeasurableImprImpr", '
                                        '"AD_PERFORMANCE_REPORT.activeViewViewableImprMeasurableImpr", '
                                        '"AD_PERFORMANCE_REPORT.allConvRate", '
                                        '"AD_PERFORMANCE_REPORT.allConvValue", '
                                        '"AD_PERFORMANCE_REPORT.avgCost", '
                                        '"AD_PERFORMANCE_REPORT.avgCPC", '
                                        '"AD_PERFORMANCE_REPORT.avgCPE", '
                                        '"AD_PERFORMANCE_REPORT.avgCPM", '
                                        '"AD_PERFORMANCE_REPORT.avgCPV", '
                                        '"AD_PERFORMANCE_REPORT.avgPosition", '
                                        '"AD_PERFORMANCE_REPORT.convRate", '
                                        '"AD_PERFORMANCE_REPORT.totalConvValue", '
                                        '"AD_PERFORMANCE_REPORT.costAllConv", '
                                        '"AD_PERFORMANCE_REPORT.costConv", '
                                        '"AD_PERFORMANCE_REPORT.costConvCurrentModel", '
                                        '"AD_PERFORMANCE_REPORT.crossDeviceConv", '
                                        '"AD_PERFORMANCE_REPORT.ctr", '
                                        '"AD_PERFORMANCE_REPORT.conversionsCurrentModel", '
                                        '"AD_PERFORMANCE_REPORT.convValueCurrentModel", '
                                        '"AD_PERFORMANCE_REPORT.engagementRate", '
                                        '"AD_PERFORMANCE_REPORT.interactionRate", '
                                        '"AD_PERFORMANCE_REPORT.interactionTypes", '
                                        '"AD_PERFORMANCE_REPORT.imprTop", '
                                        '"AD_PERFORMANCE_REPORT.valueAllConv", '
                                        '"AD_PERFORMANCE_REPORT.valueConv", '
                                        '"AD_PERFORMANCE_REPORT.valueConvCurrentModel", '
                                        '"AD_PERFORMANCE_REPORT.videoPlayedTo100", '
                                        '"AD_PERFORMANCE_REPORT.videoPlayedTo25", '
                                        '"AD_PERFORMANCE_REPORT.videoPlayedTo50", '
                                        '"AD_PERFORMANCE_REPORT.videoPlayedTo75", '
                                        '"AD_PERFORMANCE_REPORT.viewRate", '
                                        '"AD_PERFORMANCE_REPORT.viewThroughConv"]',
             'MELTANO_EXTRACT__SELECT_FILTER': '[]',
             'TAP_ADWORDS_CONVERSION_WINDOW_DAYS': '0',
             'TAP_ADWORDS_PRIMARY_KEYS': '{"KEYWORDS_PERFORMANCE_REPORT": '
                                         '["customerID", "campaignID", '
                                         '"adGroupID", "keywordID", "day", '
                                         '"network", "device"], '
                                         '"AD_PERFORMANCE_REPORT": '
                                         '["customerID", "campaignID", '
                                         '"adGroupID", "adID", "day", '
                                         '"network", "device"]}',
             'TAP_ADWORDS_USER_AGENT': 'tap-adwords via Meltano',
             'TAP_ADWORDS__LOAD_SCHEMA': 'tap_adwords',
             'TAP_ADWORDS__METADATA': '{}',
             'TAP_ADWORDS__SCHEMA': '{}',
             'TAP_ADWORDS__SELECT': '["campaigns.*", "ad_groups.*", '
                                    '"ads.*", "accounts.*", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.customerID", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.account", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.currency", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.timeZone", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.clientName", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.campaign", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.campaignID", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.campaignState", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.adGroup", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.adGroupID", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.adGroupState", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.day", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.network", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.device", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.clicks", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.cost", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.impressions", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.interactions", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.engagements", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.conversions", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.allConv", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.views", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.activeViewViewableImpressions", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.activeViewMeasurableImpr", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.activeViewMeasurableCost", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.gmailClicksToWebsite", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.gmailSaves", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.gmailForwards", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.keywordID", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.keyword", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.keywordState", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.criterionServingStatus", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.destinationURL", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.matchType", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.topOfPageCPC", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.firstPageCPC", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.imprAbsTop", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.activeViewAvgCPM", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.activeViewViewableCTR", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.activeViewMeasurableImprImpr", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.activeViewViewableImprMeasurableImpr", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.allConvRate", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.allConvValue", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.avgCost", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.avgCPC", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.avgCPE", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.avgCPM", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.avgCPV", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.avgPosition", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.convRate", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.totalConvValue", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.costAllConv", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.costConv", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.costConvCurrentModel", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.crossDeviceConv", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.ctr", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.conversionsCurrentModel", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.convValueCurrentModel", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.engagementRate", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.interactionRate", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.interactionTypes", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.imprTop", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.valueAllConv", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.valueConv", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.valueConvCurrentModel", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo100", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo25", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo50", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo75", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.viewRate", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.viewThroughConv", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.searchAbsTopIS", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.searchLostAbsTopISBudget", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.searchLostTopISBudget", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.searchExactMatchIS", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.searchImprShare", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.searchLostAbsTopISRank", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.searchLostISRank", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.searchLostTopISRank", '
                                    '"KEYWORDS_PERFORMANCE_REPORT.searchTopIS", '
                                    '"AD_PERFORMANCE_REPORT.customerID", '
                                    '"AD_PERFORMANCE_REPORT.account", '
                                    '"AD_PERFORMANCE_REPORT.currency", '
                                    '"AD_PERFORMANCE_REPORT.timeZone", '
                                    '"AD_PERFORMANCE_REPORT.clientName", '
                                    '"AD_PERFORMANCE_REPORT.campaign", '
                                    '"AD_PERFORMANCE_REPORT.campaignID", '
                                    '"AD_PERFORMANCE_REPORT.campaignState", '
                                    '"AD_PERFORMANCE_REPORT.adGroup", '
                                    '"AD_PERFORMANCE_REPORT.adGroupID", '
                                    '"AD_PERFORMANCE_REPORT.adGroupState", '
                                    '"AD_PERFORMANCE_REPORT.day", '
                                    '"AD_PERFORMANCE_REPORT.network", '
                                    '"AD_PERFORMANCE_REPORT.device", '
                                    '"AD_PERFORMANCE_REPORT.clicks", '
                                    '"AD_PERFORMANCE_REPORT.cost", '
                                    '"AD_PERFORMANCE_REPORT.impressions", '
                                    '"AD_PERFORMANCE_REPORT.interactions", '
                                    '"AD_PERFORMANCE_REPORT.engagements", '
                                    '"AD_PERFORMANCE_REPORT.conversions", '
                                    '"AD_PERFORMANCE_REPORT.allConv", '
                                    '"AD_PERFORMANCE_REPORT.views", '
                                    '"AD_PERFORMANCE_REPORT.activeViewViewableImpressions", '
                                    '"AD_PERFORMANCE_REPORT.activeViewMeasurableImpr", '
                                    '"AD_PERFORMANCE_REPORT.activeViewMeasurableCost", '
                                    '"AD_PERFORMANCE_REPORT.gmailClicksToWebsite", '
                                    '"AD_PERFORMANCE_REPORT.gmailSaves", '
                                    '"AD_PERFORMANCE_REPORT.gmailForwards", '
                                    '"AD_PERFORMANCE_REPORT.adID", '
                                    '"AD_PERFORMANCE_REPORT.adState", '
                                    '"AD_PERFORMANCE_REPORT.approvalStatus", '
                                    '"AD_PERFORMANCE_REPORT.adType", '
                                    '"AD_PERFORMANCE_REPORT.adStrength", '
                                    '"AD_PERFORMANCE_REPORT.autoAppliedAdSuggestion", '
                                    '"AD_PERFORMANCE_REPORT.ad", '
                                    '"AD_PERFORMANCE_REPORT.descriptionLine1", '
                                    '"AD_PERFORMANCE_REPORT.descriptionLine2", '
                                    '"AD_PERFORMANCE_REPORT.finalURL", '
                                    '"AD_PERFORMANCE_REPORT.displayURL", '
                                    '"AD_PERFORMANCE_REPORT.description", '
                                    '"AD_PERFORMANCE_REPORT.headline1", '
                                    '"AD_PERFORMANCE_REPORT.headline2", '
                                    '"AD_PERFORMANCE_REPORT.path1", '
                                    '"AD_PERFORMANCE_REPORT.businessName", '
                                    '"AD_PERFORMANCE_REPORT.callToActionTextResponsive", '
                                    '"AD_PERFORMANCE_REPORT.shortHeadline", '
                                    '"AD_PERFORMANCE_REPORT.longHeadline", '
                                    '"AD_PERFORMANCE_REPORT.promotionTextResponsive", '
                                    '"AD_PERFORMANCE_REPORT.responsiveSearchAdPath1", '
                                    '"AD_PERFORMANCE_REPORT.responsiveSearchAdHeadlines", '
                                    '"AD_PERFORMANCE_REPORT.responsiveSearchAdDescriptions", '
                                    '"AD_PERFORMANCE_REPORT.gmailAdBusinessName", '
                                    '"AD_PERFORMANCE_REPORT.gmailAdHeadline", '
                                    '"AD_PERFORMANCE_REPORT.gmailAdDescription", '
                                    '"AD_PERFORMANCE_REPORT.imageAdName", '
                                    '"AD_PERFORMANCE_REPORT.businessNameMultiAssetResponsiveDisplay", '
                                    '"AD_PERFORMANCE_REPORT.longHeadlineMultiAssetResponsiveDisplay", '
                                    '"AD_PERFORMANCE_REPORT.headlinesMultiAssetResponsiveDisplay", '
                                    '"AD_PERFORMANCE_REPORT.callToActionTextMultiAssetResponsiveDisplay", '
                                    '"AD_PERFORMANCE_REPORT.promotionTextMultiAssetResponsiveDisplay", '
                                    '"AD_PERFORMANCE_REPORT.imprAbsTop", '
                                    '"AD_PERFORMANCE_REPORT.activeViewAvgCPM", '
                                    '"AD_PERFORMANCE_REPORT.activeViewViewableCTR", '
                                    '"AD_PERFORMANCE_REPORT.activeViewMeasurableImprImpr", '
                                    '"AD_PERFORMANCE_REPORT.activeViewViewableImprMeasurableImpr", '
                                    '"AD_PERFORMANCE_REPORT.allConvRate", '
                                    '"AD_PERFORMANCE_REPORT.allConvValue", '
                                    '"AD_PERFORMANCE_REPORT.avgCost", '
                                    '"AD_PERFORMANCE_REPORT.avgCPC", '
                                    '"AD_PERFORMANCE_REPORT.avgCPE", '
                                    '"AD_PERFORMANCE_REPORT.avgCPM", '
                                    '"AD_PERFORMANCE_REPORT.avgCPV", '
                                    '"AD_PERFORMANCE_REPORT.avgPosition", '
                                    '"AD_PERFORMANCE_REPORT.convRate", '
                                    '"AD_PERFORMANCE_REPORT.totalConvValue", '
                                    '"AD_PERFORMANCE_REPORT.costAllConv", '
                                    '"AD_PERFORMANCE_REPORT.costConv", '
                                    '"AD_PERFORMANCE_REPORT.costConvCurrentModel", '
                                    '"AD_PERFORMANCE_REPORT.crossDeviceConv", '
                                    '"AD_PERFORMANCE_REPORT.ctr", '
                                    '"AD_PERFORMANCE_REPORT.conversionsCurrentModel", '
                                    '"AD_PERFORMANCE_REPORT.convValueCurrentModel", '
                                    '"AD_PERFORMANCE_REPORT.engagementRate", '
                                    '"AD_PERFORMANCE_REPORT.interactionRate", '
                                    '"AD_PERFORMANCE_REPORT.interactionTypes", '
                                    '"AD_PERFORMANCE_REPORT.imprTop", '
                                    '"AD_PERFORMANCE_REPORT.valueAllConv", '
                                    '"AD_PERFORMANCE_REPORT.valueConv", '
                                    '"AD_PERFORMANCE_REPORT.valueConvCurrentModel", '
                                    '"AD_PERFORMANCE_REPORT.videoPlayedTo100", '
                                    '"AD_PERFORMANCE_REPORT.videoPlayedTo25", '
                                    '"AD_PERFORMANCE_REPORT.videoPlayedTo50", '
                                    '"AD_PERFORMANCE_REPORT.videoPlayedTo75", '
                                    '"AD_PERFORMANCE_REPORT.viewRate", '
                                    '"AD_PERFORMANCE_REPORT.viewThroughConv"]',
             'TAP_ADWORDS__SELECT_FILTER': '[]'},
     'label': 'Google Ads',
     'name': 'tap-adwords',
     'namespace': 'tap_adwords',
     'pip_url': 'git+https://github.com/singer-io/tap-adwords.git',
     'plugin_type': 'extractors',
     'repo': 'https://github.com/singer-io/tap-adwords',
     'select': ['campaigns.*',
                'ad_groups.*',
                'ads.*',
                'accounts.*',
                'KEYWORDS_PERFORMANCE_REPORT.customerID',
                'KEYWORDS_PERFORMANCE_REPORT.account',
                'KEYWORDS_PERFORMANCE_REPORT.currency',
                'KEYWORDS_PERFORMANCE_REPORT.timeZone',
                'KEYWORDS_PERFORMANCE_REPORT.clientName',
                'KEYWORDS_PERFORMANCE_REPORT.campaign',
                'KEYWORDS_PERFORMANCE_REPORT.campaignID',
                'KEYWORDS_PERFORMANCE_REPORT.campaignState',
                'KEYWORDS_PERFORMANCE_REPORT.adGroup',
                'KEYWORDS_PERFORMANCE_REPORT.adGroupID',
                'KEYWORDS_PERFORMANCE_REPORT.adGroupState',
                'KEYWORDS_PERFORMANCE_REPORT.day',
                'KEYWORDS_PERFORMANCE_REPORT.network',
                'KEYWORDS_PERFORMANCE_REPORT.device',
                'KEYWORDS_PERFORMANCE_REPORT.clicks',
                'KEYWORDS_PERFORMANCE_REPORT.cost',
                'KEYWORDS_PERFORMANCE_REPORT.impressions',
                'KEYWORDS_PERFORMANCE_REPORT.interactions',
                'KEYWORDS_PERFORMANCE_REPORT.engagements',
                'KEYWORDS_PERFORMANCE_REPORT.conversions',
                'KEYWORDS_PERFORMANCE_REPORT.allConv',
                'KEYWORDS_PERFORMANCE_REPORT.views',
                'KEYWORDS_PERFORMANCE_REPORT.activeViewViewableImpressions',
                'KEYWORDS_PERFORMANCE_REPORT.activeViewMeasurableImpr',
                'KEYWORDS_PERFORMANCE_REPORT.activeViewMeasurableCost',
                'KEYWORDS_PERFORMANCE_REPORT.gmailClicksToWebsite',
                'KEYWORDS_PERFORMANCE_REPORT.gmailSaves',
                'KEYWORDS_PERFORMANCE_REPORT.gmailForwards',
                'KEYWORDS_PERFORMANCE_REPORT.keywordID',
                'KEYWORDS_PERFORMANCE_REPORT.keyword',
                'KEYWORDS_PERFORMANCE_REPORT.keywordState',
                'KEYWORDS_PERFORMANCE_REPORT.criterionServingStatus',
                'KEYWORDS_PERFORMANCE_REPORT.destinationURL',
                'KEYWORDS_PERFORMANCE_REPORT.matchType',
                'KEYWORDS_PERFORMANCE_REPORT.topOfPageCPC',
                'KEYWORDS_PERFORMANCE_REPORT.firstPageCPC',
                'KEYWORDS_PERFORMANCE_REPORT.imprAbsTop',
                'KEYWORDS_PERFORMANCE_REPORT.activeViewAvgCPM',
                'KEYWORDS_PERFORMANCE_REPORT.activeViewViewableCTR',
                'KEYWORDS_PERFORMANCE_REPORT.activeViewMeasurableImprImpr',
                'KEYWORDS_PERFORMANCE_REPORT.activeViewViewableImprMeasurableImpr',
                'KEYWORDS_PERFORMANCE_REPORT.allConvRate',
                'KEYWORDS_PERFORMANCE_REPORT.allConvValue',
                'KEYWORDS_PERFORMANCE_REPORT.avgCost',
                'KEYWORDS_PERFORMANCE_REPORT.avgCPC',
                'KEYWORDS_PERFORMANCE_REPORT.avgCPE',
                'KEYWORDS_PERFORMANCE_REPORT.avgCPM',
                'KEYWORDS_PERFORMANCE_REPORT.avgCPV',
                'KEYWORDS_PERFORMANCE_REPORT.avgPosition',
                'KEYWORDS_PERFORMANCE_REPORT.convRate',
                'KEYWORDS_PERFORMANCE_REPORT.totalConvValue',
                'KEYWORDS_PERFORMANCE_REPORT.costAllConv',
                'KEYWORDS_PERFORMANCE_REPORT.costConv',
                'KEYWORDS_PERFORMANCE_REPORT.costConvCurrentModel',
                'KEYWORDS_PERFORMANCE_REPORT.crossDeviceConv',
                'KEYWORDS_PERFORMANCE_REPORT.ctr',
                'KEYWORDS_PERFORMANCE_REPORT.conversionsCurrentModel',
                'KEYWORDS_PERFORMANCE_REPORT.convValueCurrentModel',
                'KEYWORDS_PERFORMANCE_REPORT.engagementRate',
                'KEYWORDS_PERFORMANCE_REPORT.interactionRate',
                'KEYWORDS_PERFORMANCE_REPORT.interactionTypes',
                'KEYWORDS_PERFORMANCE_REPORT.imprTop',
                'KEYWORDS_PERFORMANCE_REPORT.valueAllConv',
                'KEYWORDS_PERFORMANCE_REPORT.valueConv',
                'KEYWORDS_PERFORMANCE_REPORT.valueConvCurrentModel',
                'KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo100',
                'KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo25',
                'KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo50',
                'KEYWORDS_PERFORMANCE_REPORT.videoPlayedTo75',
                'KEYWORDS_PERFORMANCE_REPORT.viewRate',
                'KEYWORDS_PERFORMANCE_REPORT.viewThroughConv',
                'KEYWORDS_PERFORMANCE_REPORT.searchAbsTopIS',
                'KEYWORDS_PERFORMANCE_REPORT.searchLostAbsTopISBudget',
                'KEYWORDS_PERFORMANCE_REPORT.searchLostTopISBudget',
                'KEYWORDS_PERFORMANCE_REPORT.searchExactMatchIS',
                'KEYWORDS_PERFORMANCE_REPORT.searchImprShare',
                'KEYWORDS_PERFORMANCE_REPORT.searchLostAbsTopISRank',
                'KEYWORDS_PERFORMANCE_REPORT.searchLostISRank',
                'KEYWORDS_PERFORMANCE_REPORT.searchLostTopISRank',
                'KEYWORDS_PERFORMANCE_REPORT.searchTopIS',
                'AD_PERFORMANCE_REPORT.customerID',
                'AD_PERFORMANCE_REPORT.account',
                'AD_PERFORMANCE_REPORT.currency',
                'AD_PERFORMANCE_REPORT.timeZone',
                'AD_PERFORMANCE_REPORT.clientName',
                'AD_PERFORMANCE_REPORT.campaign',
                'AD_PERFORMANCE_REPORT.campaignID',
                'AD_PERFORMANCE_REPORT.campaignState',
                'AD_PERFORMANCE_REPORT.adGroup',
                'AD_PERFORMANCE_REPORT.adGroupID',
                'AD_PERFORMANCE_REPORT.adGroupState',
                'AD_PERFORMANCE_REPORT.day',
                'AD_PERFORMANCE_REPORT.network',
                'AD_PERFORMANCE_REPORT.device',
                'AD_PERFORMANCE_REPORT.clicks',
                'AD_PERFORMANCE_REPORT.cost',
                'AD_PERFORMANCE_REPORT.impressions',
                'AD_PERFORMANCE_REPORT.interactions',
                'AD_PERFORMANCE_REPORT.engagements',
                'AD_PERFORMANCE_REPORT.conversions',
                'AD_PERFORMANCE_REPORT.allConv',
                'AD_PERFORMANCE_REPORT.views',
                'AD_PERFORMANCE_REPORT.activeViewViewableImpressions',
                'AD_PERFORMANCE_REPORT.activeViewMeasurableImpr',
                'AD_PERFORMANCE_REPORT.activeViewMeasurableCost',
                'AD_PERFORMANCE_REPORT.gmailClicksToWebsite',
                'AD_PERFORMANCE_REPORT.gmailSaves',
                'AD_PERFORMANCE_REPORT.gmailForwards',
                'AD_PERFORMANCE_REPORT.adID',
                'AD_PERFORMANCE_REPORT.adState',
                'AD_PERFORMANCE_REPORT.approvalStatus',
                'AD_PERFORMANCE_REPORT.adType',
                'AD_PERFORMANCE_REPORT.adStrength',
                'AD_PERFORMANCE_REPORT.autoAppliedAdSuggestion',
                'AD_PERFORMANCE_REPORT.ad',
                'AD_PERFORMANCE_REPORT.descriptionLine1',
                'AD_PERFORMANCE_REPORT.descriptionLine2',
                'AD_PERFORMANCE_REPORT.finalURL',
                'AD_PERFORMANCE_REPORT.displayURL',
                'AD_PERFORMANCE_REPORT.description',
                'AD_PERFORMANCE_REPORT.headline1',
                'AD_PERFORMANCE_REPORT.headline2',
                'AD_PERFORMANCE_REPORT.path1',
                'AD_PERFORMANCE_REPORT.businessName',
                'AD_PERFORMANCE_REPORT.callToActionTextResponsive',
                'AD_PERFORMANCE_REPORT.shortHeadline',
                'AD_PERFORMANCE_REPORT.longHeadline',
                'AD_PERFORMANCE_REPORT.promotionTextResponsive',
                'AD_PERFORMANCE_REPORT.responsiveSearchAdPath1',
                'AD_PERFORMANCE_REPORT.responsiveSearchAdHeadlines',
                'AD_PERFORMANCE_REPORT.responsiveSearchAdDescriptions',
                'AD_PERFORMANCE_REPORT.gmailAdBusinessName',
                'AD_PERFORMANCE_REPORT.gmailAdHeadline',
                'AD_PERFORMANCE_REPORT.gmailAdDescription',
                'AD_PERFORMANCE_REPORT.imageAdName',
                'AD_PERFORMANCE_REPORT.businessNameMultiAssetResponsiveDisplay',
                'AD_PERFORMANCE_REPORT.longHeadlineMultiAssetResponsiveDisplay',
                'AD_PERFORMANCE_REPORT.headlinesMultiAssetResponsiveDisplay',
                'AD_PERFORMANCE_REPORT.callToActionTextMultiAssetResponsiveDisplay',
                'AD_PERFORMANCE_REPORT.promotionTextMultiAssetResponsiveDisplay',
                'AD_PERFORMANCE_REPORT.imprAbsTop',
                'AD_PERFORMANCE_REPORT.activeViewAvgCPM',
                'AD_PERFORMANCE_REPORT.activeViewViewableCTR',
                'AD_PERFORMANCE_REPORT.activeViewMeasurableImprImpr',
                'AD_PERFORMANCE_REPORT.activeViewViewableImprMeasurableImpr',
                'AD_PERFORMANCE_REPORT.allConvRate',
                'AD_PERFORMANCE_REPORT.allConvValue',
                'AD_PERFORMANCE_REPORT.avgCost',
                'AD_PERFORMANCE_REPORT.avgCPC',
                'AD_PERFORMANCE_REPORT.avgCPE',
                'AD_PERFORMANCE_REPORT.avgCPM',
                'AD_PERFORMANCE_REPORT.avgCPV',
                'AD_PERFORMANCE_REPORT.avgPosition',
                'AD_PERFORMANCE_REPORT.convRate',
                'AD_PERFORMANCE_REPORT.totalConvValue',
                'AD_PERFORMANCE_REPORT.costAllConv',
                'AD_PERFORMANCE_REPORT.costConv',
                'AD_PERFORMANCE_REPORT.costConvCurrentModel',
                'AD_PERFORMANCE_REPORT.crossDeviceConv',
                'AD_PERFORMANCE_REPORT.ctr',
                'AD_PERFORMANCE_REPORT.conversionsCurrentModel',
                'AD_PERFORMANCE_REPORT.convValueCurrentModel',
                'AD_PERFORMANCE_REPORT.engagementRate',
                'AD_PERFORMANCE_REPORT.interactionRate',
                'AD_PERFORMANCE_REPORT.interactionTypes',
                'AD_PERFORMANCE_REPORT.imprTop',
                'AD_PERFORMANCE_REPORT.valueAllConv',
                'AD_PERFORMANCE_REPORT.valueConv',
                'AD_PERFORMANCE_REPORT.valueConvCurrentModel',
                'AD_PERFORMANCE_REPORT.videoPlayedTo100',
                'AD_PERFORMANCE_REPORT.videoPlayedTo25',
                'AD_PERFORMANCE_REPORT.videoPlayedTo50',
                'AD_PERFORMANCE_REPORT.videoPlayedTo75',
                'AD_PERFORMANCE_REPORT.viewRate',
                'AD_PERFORMANCE_REPORT.viewThroughConv'],
     'settings': [{'description': 'Your Developer Token for Google AdWord '
                                  'Application',
                   'kind': 'password',
                   'label': 'Developer Token',
                   'name': 'developer_token',
                   'placeholder': 'Ex. *****************'},
                  {'description': 'Your Google OAuth Client ID',
                   'kind': 'password',
                   'label': 'OAuth Client ID',
                   'name': 'oauth_client_id',
                   'placeholder': 'Ex. '
                                  '123456789012345.apps.googleusercontent.com'},
                  {'description': 'Your Google OAuth Client Secret',
                   'kind': 'password',
                   'label': 'OAuth Client Secret',
                   'name': 'oauth_client_secret',
                   'placeholder': 'Ex. *****************'},
                  {'description': 'The Refresh Token generated through the '
                                  'OAuth flow run using your OAuth Client '
                                  'and your Developer Token',
                   'kind': 'oauth',
                   'label': 'Access Token',
                   'name': 'refresh_token',
                   'oauth': {'provider': 'google-adwords'},
                   'placeholder': 'Ex. *****************'},
                  {'description': 'A comma-separated list of Ad Account '
                                  'IDs to replicate data from',
                   'label': 'Account ID(s)',
                   'name': 'customer_ids',
                   'placeholder': 'Ex. 1234567890,1234567891,1234567892'},
                  {'description': 'Determines how much historical data '
                                  'will be extracted. Please be aware that '
                                  'the larger the time period and amount '
                                  'of data, the longer the initial '
                                  'extraction can be expected to take.',
                   'kind': 'date_iso8601',
                   'name': 'start_date'},
                  {'description': 'Date up to when historical data will be '
                                  'extracted.',
                   'kind': 'date_iso8601',
                   'name': 'end_date'},
                  {'description': 'The User Agent for your OAuth Client '
                                  '(used in requests made to the AdWords '
                                  'API)',
                   'label': 'User Agent for your OAuth Client',
                   'name': 'user_agent',
                   'placeholder': 'Ex. tap-adwords via Meltano '
                                  '<user@example.com>',
                   'value': 'tap-adwords via Meltano'},
                  {'description': 'How many Days before the Start Date to '
                                  'fetch data for Performance Reports',
                   'kind': 'integer',
                   'label': 'Conversion Window Days',
                   'name': 'conversion_window_days',
                   'value': 0},
                  {'description': 'Primary Keys for the selected Entities '
                                  '(Streams)',
                   'kind': 'object',
                   'label': 'Primary Keys',
                   'name': 'primary_keys',
                   'value': {'AD_PERFORMANCE_REPORT': ['customerID',
                                                       'campaignID',
                                                       'adGroupID',
                                                       'adID',
                                                       'day',
                                                       'network',
                                                       'device'],
                             'KEYWORDS_PERFORMANCE_REPORT': ['customerID',
                                                             'campaignID',
                                                             'adGroupID',
                                                             'keywordID',
                                                             'day',
                                                             'network',
                                                             'device']}}],
     'settings_group_validation': [['developer_token',
                                    'oauth_client_id',
                                    'oauth_client_secret',
                                    'refresh_token',
                                    'user_agent',
                                    'customer_ids',
                                    'start_date']],
     'variant': 'singer-io'}
Unevaluated properties are not allowed ('plugin_type', 'repo' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['extractors']['items']:
    {'allOf': [{'$ref': '#/$defs/plugins/plugin_generic'},
               {'$ref': '#/$defs/plugins/extractor_shared'},
               {'properties': {'capabilities': {'items': {'enum': ['properties',
                                                                   'catalog',
                                                                   'discover',
                                                                   'state',
                                                                   'about',
                                                                   'stream-maps',
                                                                   'activate-version',
                                                                   'batch',
                                                                   'test',
                                                                   'log-based'],
                                                          'type': 'string'},
                                                'type': 'array'}},
                'type': 'object'}],
     'description': 'https://docs.meltano.com/concepts/plugins#extractors',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['extractors'][2]:
    {'capabilities': ['catalog', 'discover', 'state'],
     'docs': 'https://hub.meltano.com/extractors/google-analytics.html',
     'env': {'MELTANO_EXTRACT_KEY_FILE_LOCATION': '/home/will/meltano/projects/sandbox/client_secrets.json',
             'MELTANO_EXTRACT__LOAD_SCHEMA': 'tap_google_analytics',
             'MELTANO_EXTRACT__METADATA': '{}',
             'MELTANO_EXTRACT__SCHEMA': '{}',
             'MELTANO_EXTRACT__SELECT': '["*.*"]',
             'MELTANO_EXTRACT__SELECT_FILTER': '[]',
             'TAP_GOOGLE_ANALYTICS_KEY_FILE_LOCATION': '/home/will/meltano/projects/sandbox/client_secrets.json',
             'TAP_GOOGLE_ANALYTICS__LOAD_SCHEMA': 'tap_google_analytics',
             'TAP_GOOGLE_ANALYTICS__METADATA': '{}',
             'TAP_GOOGLE_ANALYTICS__SCHEMA': '{}',
             'TAP_GOOGLE_ANALYTICS__SELECT': '["*.*"]',
             'TAP_GOOGLE_ANALYTICS__SELECT_FILTER': '[]'},
     'label': 'Google Analytics',
     'name': 'tap-google-analytics',
     'namespace': 'tap_google_analytics',
     'pip_url': 'git+https://github.com/MeltanoLabs/tap-google-analytics.git',
     'plugin_type': 'extractors',
     'repo': 'https://github.com/MeltanoLabs/tap-google-analytics',
     'settings': [{'description': 'A file that contains the Google '
                                  'Analytics client secrets json.',
                   'kind': 'file',
                   'label': 'Client Secrets File Location',
                   'name': 'key_file_location',
                   'placeholder': 'Ex. client_secrets.json',
                   'value': '$MELTANO_PROJECT_ROOT/client_secrets.json'},
                  {'description': 'An object that contains the Google '
                                  'Analytics client secrets.',
                   'kind': 'object',
                   'label': 'Client Secrets JSON',
                   'name': 'client_secrets',
                   'placeholder': 'Ex. client_secrets.json'},
                  {'description': 'The Google Analytics oauth client ID.',
                   'kind': 'password',
                   'label': 'OAuth Client ID',
                   'name': 'oauth_credentials.client_id'},
                  {'description': 'The Google Analytics oauth client '
                                  'secret.',
                   'kind': 'password',
                   'label': 'OAuth Client Secret',
                   'name': 'oauth_credentials.client_secret'},
                  {'description': 'The Google Analytics oauth access '
                                  'token.',
                   'kind': 'password',
                   'label': 'OAuth Access Token',
                   'name': 'oauth_credentials.access_token'},
                  {'description': 'The Google Analytics oauth refresh '
                                  'token.',
                   'kind': 'password',
                   'label': 'OAuth Refresh Token',
                   'name': 'oauth_credentials.refresh_token'},
                  {'description': 'The ID for the view to fetch data from.',
                   'label': 'View ID',
                   'name': 'view_id',
                   'placeholder': 'Ex. 198343027'},
                  {'description': 'The reports definition of which fields '
                                  'to retrieve from the view.',
                   'label': 'Reports',
                   'name': 'reports',
                   'placeholder': 'Ex. my_report_definition.json'},
                  {'description': 'This property determines how much '
                                  'historical data will be extracted. '
                                  'Please be aware that the larger the '
                                  'time period and amount of data, the '
                                  'longer the initial extraction can be '
                                  'expected to take.',
                   'kind': 'date_iso8601',
                   'label': 'Start Date',
                   'name': 'start_date'},
                  {'description': 'Date up to when historical data will be '
                                  'extracted.',
                   'kind': 'date_iso8601',
                   'label': 'End Date',
                   'name': 'end_date'}],
     'settings_group_validation': [['key_file_location',
                                    'view_id',
                                    'start_date'],
                                   ['client_secrets',
                                    'view_id',
                                    'start_date'],
                                   ['oauth_credentials.client_id',
                                    'oauth_credentials.client_secret',
                                    'oauth_credentials.access_token',
                                    'oauth_credentials.refresh_token',
                                    'view_id',
                                    'start_date']],
     'variant': 'meltanolabs'}
Unevaluated properties are not allowed ('description', 'plugin_type', 'repo' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['loaders']['items']:
    {'allOf': [{'$ref': '#/$defs/plugins/plugin_generic'},
               {'properties': {'capabilities': {'items': {'enum': ['about',
                                                                   'stream-maps',
                                                                   'activate-version',
                                                                   'batch',
                                                                   'soft-delete',
                                                                   'hard-delete',
                                                                   'datatype-failsafe',
                                                                   'record-flattening'],
                                                          'type': 'string'},
                                                'type': 'array'},
                               'dialect': {'description': 'The name of the '
                                                          'dialect of the '
                                                          'target '
                                                          'database, so '
                                                          'that '
                                                          'transformers in '
                                                          'the same '
                                                          'pipeline and '
                                                          "Meltano UI's "
                                                          'Analysis '
                                                          'feature can '
                                                          'determine the '
                                                          'type of '
                                                          'database to '
                                                          'connect to.',
                                           'type': 'string'},
                               'target_schema': {'description': 'The name '
                                                                'of the '
                                                                'database '
                                                                'schema '
                                                                'the '
                                                                'loader '
                                                                'has been '
                                                                'configured '
                                                                'to load '
                                                                'data into '
                                                                '(assuming '
                                                                'the '
                                                                'destination '
                                                                'supports '
                                                                'schemas), '
                                                                'so that '
                                                                'transformers '
                                                                'in the '
                                                                'same '
                                                                'pipeline '
                                                                'can '
                                                                'determine '
                                                                'the '
                                                                'database '
                                                                'schema to '
                                                                'load data '
                                                                'from.',
                                                 'type': 'string'}}}],
     'description': 'https://docs.meltano.com/concepts/plugins#loaders',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['loaders'][0]:
    {'config': {'destination_path': '${DST_PATH}/dir'},
     'description': 'JSONL loader',
     'docs': 'https://hub.meltano.com/loaders/target-jsonl--andyh1203',
     'env': {'MELTANO_LOAD_DESTINATION_PATH': '/dir',
             'MELTANO_LOAD_DO_TIMESTAMP_FILE': 'false',
             'MELTANO_LOAD__DIALECT': 'target_jsonl',
             'TARGET_JSONL_DESTINATION_PATH': '/dir',
             'TARGET_JSONL_DO_TIMESTAMP_FILE': 'false',
             'TARGET_JSONL__DIALECT': 'target_jsonl'},
     'label': 'JSON Lines (JSONL)',
     'logo_url': 'https://hub.meltano.com/assets/logos/loaders/jsonl.png',
     'name': 'target-jsonl',
     'namespace': 'target_jsonl',
     'pip_url': 'target-jsonl',
     'plugin_type': 'loaders',
     'repo': 'https://github.com/andyh1203/target-jsonl',
     'settings': [{'description': 'Sets the destination path the JSONL '
                                  'files are written to, relative\n'
                                  'to the project root.\n'
                                  '\n'
                                  'The directory needs to exist already, '
                                  'it will not be created\n'
                                  'automatically.\n'
                                  '\n'
                                  'To write JSONL files to the project '
                                  'root, set an empty string (`""`).\n',
                   'kind': 'string',
                   'label': 'Destination Path',
                   'name': 'destination_path',
                   'value': 'output'},
                  {'description': 'Specifies if the files should get '
                                  'timestamped.\n'
                                  '\n'
                                  'By default, the resulting file will not '
                                  'have a timestamp in the file name (i.e. '
                                  '`exchange_rate.jsonl`).\n'
                                  '\n'
                                  'If this option gets set to `true`, the '
                                  'resulting file will have a timestamp '
                                  'associated with it (i.e. '
                                  '`exchange_rate-{timestamp}.jsonl`).\n',
                   'kind': 'boolean',
                   'label': 'Include Timestamp in File Names',
                   'name': 'do_timestamp_file',
                   'value': False},
                  {'description': 'Specifies a custom name for the '
                                  'filename, instead of the stream name.\n'
                                  '\n'
                                  'The file name will be '
                                  '`{custom_name}-{timestamp}.jsonl`, if '
                                  '`do_timestamp_file` is `true`.\n'
                                  'Otherwise the file name will be '
                                  '`{custom_name}.jsonl`.\n'
                                  '\n'
                                  'If custom name is not provided, the '
                                  'stream name will be used.\n',
                   'kind': 'string',
                   'label': 'Custom File Name Override',
                   'name': 'custom_name'}],
     'variant': 'andyh1203'}
Additional properties are not allowed ('value_post_processor' was unexpected)

Failed validating 'additionalProperties' in schema['properties']['plugins']['properties']['loaders']['items']['allOf'][0]['properties']['settings']['items']:
    {'additionalProperties': False,
     'properties': {'aliases': {'description': 'Alternative setting names '
                                               'that can be used in '
                                               "'meltano.yml' and 'meltano "
                                               "config set'",
                                'examples': ['accountId',
                                             'account_identifier'],
                                'items': {'type': 'string'},
                                'type': 'array'},
                    'description': {'description': 'A description for what '
                                                   'this setting does',
                                    'examples': ['The unique account '
                                                 'identifier for your '
                                                 'Stripe Account'],
                                    'type': 'string'},
                    'documentation': {'description': 'A link to '
                                                     'documentation on '
                                                     'this setting',
                                      'examples': ['https://meltano.com/'],
                                      'type': 'string'},
                    'env': {'description': 'An alternative environment '
                                           'variable name to populate with '
                                           'this settings value in the '
                                           'plugin environment.',
                            'examples': ['GITLAB_API_TOKEN',
                                         'FACEBOOK_ADS_ACCESS_TOKEN'],
                            'type': 'string'},
                    'kind': {'description': 'The type of value this '
                                            'setting contains',
                             'enum': ['oauth',
                                      'hidden',
                                      'password',
                                      'date_iso8601',
                                      'file',
                                      'email',
                                      'integer',
                                      'options',
                                      'object',
                                      'array',
                                      'boolean',
                                      'string'],
                             'type': 'string'},
                    'label': {'description': 'A user friendly label for '
                                             'the setting',
                              'examples': ['Account ID'],
                              'type': 'string'},
                    'name': {'description': 'The name of the setting',
                             'examples': ['account_id'],
                             'type': 'string'},
                    'oauth': {'properties': {'provider': {'description': 'The '
                                                                         'name '
                                                                         'of '
                                                                         'a '
                                                                         'Meltano-supported '
                                                                         'OAuth '
                                                                         'provider',
                                                          'examples': ['google-adwords'],
                                                          'type': 'string'}},
                              'required': ['provider'],
                              'type': 'object'},
                    'placeholder': {'description': 'A placeholder value '
                                                   'for this setting',
                                    'examples': ['Ex. 18161'],
                                    'type': 'string'},
                    'protected': {'default': False,
                                  'description': 'A protected setting '
                                                 'cannot be changed from '
                                                 'the UI',
                                  'type': 'boolean'},
                    'tooltip': {'description': 'A phrase to provide '
                                               'additional information on '
                                               'this setting',
                                'examples': ['Here is some additional '
                                             'info...'],
                                'type': 'string'},
                    'value': {'description': 'The default value of this '
                                             'setting if not otherwise '
                                             'defined'},
                    'value_processor': {'const': 'nest_object',
                                        'description': 'Use with `kind: '
                                                       'object` to '
                                                       'automatically nest '
                                                       'flattened keys'}},
     'required': ['name'],
     'type': 'object'}

On instance['plugins']['loaders'][1]['settings'][5]:
    {'description': "Using SSL via postgres `sslmode='require'` option.\n"
                    '\n'
                    'If the server does not accept SSL connections or the '
                    'client certificate is not recognized the connection '
                    'will fail.\n',
     'kind': 'boolean',
     'label': 'SSL',
     'name': 'ssl',
     'value': False,
     'value_post_processor': 'stringify'}
Unevaluated properties are not allowed ('docs', 'env', 'label', 'name', 'namespace', 'pip_url', 'plugin_type', 'repo', 'settings', 'settings_group_validation', 'variant' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['loaders']['items']:
    {'allOf': [{'$ref': '#/$defs/plugins/plugin_generic'},
               {'properties': {'capabilities': {'items': {'enum': ['about',
                                                                   'stream-maps',
                                                                   'activate-version',
                                                                   'batch',
                                                                   'soft-delete',
                                                                   'hard-delete',
                                                                   'datatype-failsafe',
                                                                   'record-flattening'],
                                                          'type': 'string'},
                                                'type': 'array'},
                               'dialect': {'description': 'The name of the '
                                                          'dialect of the '
                                                          'target '
                                                          'database, so '
                                                          'that '
                                                          'transformers in '
                                                          'the same '
                                                          'pipeline and '
                                                          "Meltano UI's "
                                                          'Analysis '
                                                          'feature can '
                                                          'determine the '
                                                          'type of '
                                                          'database to '
                                                          'connect to.',
                                           'type': 'string'},
                               'target_schema': {'description': 'The name '
                                                                'of the '
                                                                'database '
                                                                'schema '
                                                                'the '
                                                                'loader '
                                                                'has been '
                                                                'configured '
                                                                'to load '
                                                                'data into '
                                                                '(assuming '
                                                                'the '
                                                                'destination '
                                                                'supports '
                                                                'schemas), '
                                                                'so that '
                                                                'transformers '
                                                                'in the '
                                                                'same '
                                                                'pipeline '
                                                                'can '
                                                                'determine '
                                                                'the '
                                                                'database '
                                                                'schema to '
                                                                'load data '
                                                                'from.',
                                                 'type': 'string'}}}],
     'description': 'https://docs.meltano.com/concepts/plugins#loaders',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['loaders'][1]:
    {'dialect': 'postgres',
     'docs': 'https://hub.meltano.com/loaders/target-postgres--transferwise',
     'env': {'MELTANO_LOAD_ADD_METADATA_COLUMNS': 'false',
             'MELTANO_LOAD_BATCH_SIZE_ROWS': '100000',
             'MELTANO_LOAD_DATA_FLATTENING_MAX_LEVEL': '0',
             'MELTANO_LOAD_FLUSH_ALL_STREAMS': 'false',
             'MELTANO_LOAD_HARD_DELETE': 'false',
             'MELTANO_LOAD_HOST': 'localhost',
             'MELTANO_LOAD_PARALLELISM': '0',
             'MELTANO_LOAD_PARALLELISM_MAX': '16',
             'MELTANO_LOAD_PORT': '5432',
             'MELTANO_LOAD_PRIMARY_KEY_REQUIRED': 'true',
             'MELTANO_LOAD_SSL': 'false',
             'MELTANO_LOAD_VALIDATE_RECORDS': 'false',
             'MELTANO_LOAD__DIALECT': 'postgres',
             'TARGET_POSTGRES_ADD_METADATA_COLUMNS': 'false',
             'TARGET_POSTGRES_BATCH_SIZE_ROWS': '100000',
             'TARGET_POSTGRES_DATA_FLATTENING_MAX_LEVEL': '0',
             'TARGET_POSTGRES_FLUSH_ALL_STREAMS': 'false',
             'TARGET_POSTGRES_HARD_DELETE': 'false',
             'TARGET_POSTGRES_HOST': 'localhost',
             'TARGET_POSTGRES_PARALLELISM': '0',
             'TARGET_POSTGRES_PARALLELISM_MAX': '16',
             'TARGET_POSTGRES_PORT': '5432',
             'TARGET_POSTGRES_PRIMARY_KEY_REQUIRED': 'true',
             'TARGET_POSTGRES_SSL': 'false',
             'TARGET_POSTGRES_VALIDATE_RECORDS': 'false',
             'TARGET_POSTGRES__DIALECT': 'postgres'},
     'label': 'PostgreSQL',
     'name': 'target-postgres',
     'namespace': 'target_postgres',
     'pip_url': 'pipelinewise-target-postgres',
     'plugin_type': 'loaders',
     'repo': 'https://github.com/transferwise/pipelinewise-target-postgres',
     'settings': [{'description': 'PostgreSQL host',
                   'label': 'Host',
                   'name': 'host',
                   'value': 'localhost'},
                  {'description': 'PostgreSQL port',
                   'kind': 'integer',
                   'label': 'Port',
                   'name': 'port',
                   'value': 5432},
                  {'description': 'PostgreSQL user',
                   'label': 'User',
                   'name': 'user'},
                  {'description': 'PostgreSQL password',
                   'kind': 'password',
                   'label': 'Password',
                   'name': 'password'},
                  {'description': 'PostgreSQL database name',
                   'label': 'Database Name',
                   'name': 'dbname'},
                  {'description': 'Using SSL via postgres '
                                  "`sslmode='require'` option.\n"
                                  '\n'
                                  'If the server does not accept SSL '
                                  'connections or the client certificate '
                                  'is not recognized the connection will '
                                  'fail.\n',
                   'kind': 'boolean',
                   'label': 'SSL',
                   'name': 'ssl',
                   'value': False,
                   'value_post_processor': 'stringify'},
                  {'aliases': ['schema'],
                   'description': 'Note that '
                                  '`$MELTANO_EXTRACT__LOAD_SCHEMA` [will '
                                  'expand '
                                  'to](https://docs.meltano.com/guide/configuration.html#expansion-in-setting-values) '
                                  'the value of the [`load_schema` '
                                  'extra](https://docs.meltano.com/concepts/plugins#load-schema-extra) '
                                  'for the extractor used in the pipeline, '
                                  "which defaults to the extractor's "
                                  'namespace, e.g. `tap_gitlab` for '
                                  '[`tap-gitlab`](/extractors/gitlab).\n'
                                  '\n'
                                  'Name of the schema where the tables '
                                  'will be created. If `schema_mapping`\n'
                                  'is not defined then every stream sent '
                                  'by the tap is loaded into this '
                                  'schema.\n',
                   'label': 'Default Target Schema',
                   'name': 'default_target_schema',
                   'value': '$MELTANO_EXTRACT__LOAD_SCHEMA'},
                  {'description': 'Maximum number of rows in each batch. '
                                  'At the end of each batch, the rows in '
                                  'the batch are loaded into Postgres.',
                   'kind': 'integer',
                   'label': 'Batch Size Rows',
                   'name': 'batch_size_rows',
                   'value': 100000},
                  {'description': 'Flush and load every stream into '
                                  'Postgres when one batch is full. '
                                  'Warning: This may trigger the COPY '
                                  'command to use files with low number of '
                                  'records.',
                   'kind': 'boolean',
                   'label': 'Flush All Streams',
                   'name': 'flush_all_streams',
                   'value': False},
                  {'description': 'The number of threads used to flush '
                                  'tables. 0 will create a thread for each '
                                  'stream, up to parallelism_max. -1 will '
                                  'create a thread for each CPU core. Any '
                                  'other positive number will create that '
                                  'number of threads, up to '
                                  'parallelism_max.',
                   'kind': 'integer',
                   'label': 'Parallelism',
                   'name': 'parallelism',
                   'value': 0},
                  {'description': 'Max number of parallel threads to use '
                                  'when flushing tables.',
                   'kind': 'integer',
                   'label': 'Max Parallelism',
                   'name': 'parallelism_max',
                   'value': 16},
                  {'description': 'Grant USAGE privilege on newly created '
                                  'schemas and grant SELECT privilege on '
                                  'newly created tables to a specific role '
                                  'or a list of roles. If `schema_mapping` '
                                  'is not defined then every stream sent '
                                  'by the tap is granted accordingly.',
                   'label': 'Default Target Schema Select Permission',
                   'name': 'default_target_schema_select_permission'},
                  {'description': 'Useful if you want to load multiple '
                                  'streams from one tap to multiple '
                                  'Postgres schemas.\n'
                                  'If the tap sends the `stream_id` in '
                                  '`<schema_name>-<table_name>` format '
                                  'then this option overwrites the '
                                  '`default_target_schema` value. Note, '
                                  'that using `schema_mapping` you can '
                                  'overwrite the '
                                  '`default_target_schema_select_permission` '
                                  'value to grant SELECT permissions to '
                                  'different groups per schemas or '
                                  'optionally you can create indices '
                                  'automatically for the replicated '
                                  'tables.\n',
                   'kind': 'object',
                   'label': 'Schema Mapping',
                   'name': 'schema_mapping'},
                  {'description': 'Useful if you want to load multiple '
                                  'streams from one tap to multiple '
                                  'Postgres schemas.\n'
                                  '\n'
                                  'If the tap sends the `stream_id` in '
                                  '`<schema_name>-<table_name>` format '
                                  'then this option overwrites the '
                                  '`default_target_schema` value.\n'
                                  '\n'
                                  'Note, that using `schema_mapping` you '
                                  'can overwrite the '
                                  '`default_target_schema_select_permission` '
                                  'value to grant SELECT permissions to '
                                  'different groups per schemas or '
                                  'optionally you can create indices '
                                  'automatically for the replicated '
                                  'tables.\n'
                                  '\n'
                                  'This setting can hold an object mapping '
                                  'source schema names to objects with '
                                  '`target_schema` and (optionally) '
                                  '`target_schema_select_permissions` and '
                                  '`indices` keys.\n',
                   'kind': 'boolean',
                   'label': 'Add Metadata Columns',
                   'name': 'add_metadata_columns',
                   'value': False},
                  {'description': 'When `hard_delete` option is true then '
                                  'DELETE SQL commands will be performed '
                                  'in Postgres to delete rows in tables. '
                                  "It's achieved by continuously checking "
                                  'the `_SDC_DELETED_AT` metadata column '
                                  'sent by the singer tap. Due to deleting '
                                  'rows requires metadata columns, '
                                  '`hard_delete` option automatically '
                                  'enables the `add_metadata_columns` '
                                  'option as well.',
                   'kind': 'boolean',
                   'label': 'Hard Delete',
                   'name': 'hard_delete',
                   'value': False},
                  {'description': 'Object type `RECORD`` items from taps '
                                  'can be transformed to flattened columns '
                                  'by creating columns automatically. When '
                                  'value is 0 (default) then flattening '
                                  'functionality is turned off.',
                   'kind': 'integer',
                   'label': 'Data Flattening Max Level',
                   'name': 'data_flattening_max_level',
                   'value': 0},
                  {'description': 'Log based and Incremental replications '
                                  'on tables with no Primary Key cause '
                                  'duplicates when merging UPDATE events. '
                                  'When set to true, stop loading data if '
                                  'no Primary Key is defined.',
                   'kind': 'boolean',
                   'label': 'Primary Key Required',
                   'name': 'primary_key_required',
                   'value': True},
                  {'description': 'Validate every single record message to '
                                  'the corresponding JSON schema. This '
                                  'option is disabled by default and '
                                  'invalid RECORD messages will fail only '
                                  'at load time by Postgres. Enabling this '
                                  'option will detect invalid records '
                                  'earlier but could cause performance '
                                  'degradation.',
                   'kind': 'boolean',
                   'label': 'Validate Records',
                   'name': 'validate_records',
                   'value': False},
                  {'description': '(Default: platform-dependent) Directory '
                                  'of temporary CSV files with RECORD '
                                  'messages.',
                   'label': 'Temporary Directory',
                   'name': 'temp_dir'}],
     'settings_group_validation': [['host',
                                    'port',
                                    'user',
                                    'password',
                                    'dbname',
                                    'default_target_schema']],
     'target_schema': '$TARGET_POSTGRES_DEFAULT_TARGET_SCHEMA',
     'variant': 'transferwise'}
Unevaluated properties are not allowed ('plugin_type', 'repo' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['orchestrators']['items']:
    {'$ref': '#/$defs/plugins/plugin_generic',
     'description': 'https://docs.meltano.com/concepts/plugins#orchestrators',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['orchestrators'][0]:
    {'commands': {'create-admin': {'args': 'users create --username admin '
                                           '--firstname FIRST_NAME '
                                           '--lastname LAST_NAME --role '
                                           'Admin --email '
                                           'admin@example.org',
                                   'description': 'Create an admin user.'},
                  'ui': {'args': 'webserver',
                         'description': 'Start the Airflow webserver.'}},
     'docs': 'https://docs.meltano.com/guide/orchestration',
     'env': {'AIRFLOW_CORE_DAGS_ARE_PAUSED_AT_CREATION': 'False',
             'AIRFLOW_CORE_DAGS_FOLDER': '/home/will/meltano/projects/sandbox/orchestrate/dags',
             'AIRFLOW_CORE_LOAD_EXAMPLES': 'False',
             'AIRFLOW_CORE_PLUGINS_FOLDER': '/home/will/meltano/projects/sandbox/orchestrate/plugins',
             'AIRFLOW_CORE_SQL_ALCHEMY_CONN': 'sqlite:////home/will/meltano/projects/sandbox/.meltano/orchestrators/airflow/airflow.db',
             'AIRFLOW_WEBSERVER_WEB_SERVER_PORT': '8080',
             'AIRFLOW__CORE__DAGS_ARE_PAUSED_AT_CREATION': 'False',
             'AIRFLOW__CORE__DAGS_FOLDER': '/home/will/meltano/projects/sandbox/orchestrate/dags',
             'AIRFLOW__CORE__LOAD_EXAMPLES': 'False',
             'AIRFLOW__CORE__PLUGINS_FOLDER': '/home/will/meltano/projects/sandbox/orchestrate/plugins',
             'AIRFLOW__CORE__SQL_ALCHEMY_CONN': 'sqlite:////home/will/meltano/projects/sandbox/.meltano/orchestrators/airflow/airflow.db',
             'AIRFLOW__WEBSERVER__WEB_SERVER_PORT': '8080',
             'MELTANO_ORCHESTRAT_CORE_DAGS_ARE_PAUSED_AT_CREATION': 'False',
             'MELTANO_ORCHESTRAT_CORE_DAGS_FOLDER': '/home/will/meltano/projects/sandbox/orchestrate/dags',
             'MELTANO_ORCHESTRAT_CORE_LOAD_EXAMPLES': 'False',
             'MELTANO_ORCHESTRAT_CORE_PLUGINS_FOLDER': '/home/will/meltano/projects/sandbox/orchestrate/plugins',
             'MELTANO_ORCHESTRAT_CORE_SQL_ALCHEMY_CONN': 'sqlite:////home/will/meltano/projects/sandbox/.meltano/orchestrators/airflow/airflow.db',
             'MELTANO_ORCHESTRAT_WEBSERVER_WEB_SERVER_PORT': '8080'},
     'label': 'Airflow',
     'name': 'airflow',
     'namespace': 'airflow',
     'pip_url': 'apache-airflow==2.1.2 --constraint '
                'https://raw.githubusercontent.com/apache/airflow/constraints-2.1.2/constraints-${MELTANO__PYTHON_VERSION}.txt',
     'plugin_type': 'orchestrators',
     'repo': 'https://github.com/apache/airflow',
     'requires': {'files': [{'name': 'files-airflow',
                             'variant': 'meltano'}]},
     'settings': [{'env': 'AIRFLOW__CORE__DAGS_FOLDER',
                   'label': 'DAGs Folder',
                   'name': 'core.dags_folder',
                   'value': '$MELTANO_PROJECT_ROOT/orchestrate/dags'},
                  {'env': 'AIRFLOW__CORE__PLUGINS_FOLDER',
                   'label': 'Plugins Folder',
                   'name': 'core.plugins_folder',
                   'value': '$MELTANO_PROJECT_ROOT/orchestrate/plugins'},
                  {'env': 'AIRFLOW__CORE__SQL_ALCHEMY_CONN',
                   'label': 'SQL Alchemy Connection',
                   'name': 'core.sql_alchemy_conn',
                   'value': 'sqlite:///$MELTANO_PROJECT_ROOT/.meltano/orchestrators/airflow/airflow.db'},
                  {'env': 'AIRFLOW__CORE__LOAD_EXAMPLES',
                   'label': 'Load Examples',
                   'name': 'core.load_examples',
                   'value': False},
                  {'env': 'AIRFLOW__CORE__DAGS_ARE_PAUSED_AT_CREATION',
                   'label': 'Pause DAGs at Creation',
                   'name': 'core.dags_are_paused_at_creation',
                   'value': False},
                  {'env': 'AIRFLOW__WEBSERVER__WEB_SERVER_PORT',
                   'label': 'Webserver Port',
                   'name': 'webserver.web_server_port',
                   'value': 8080}],
     'variant': 'apache'}
Unevaluated properties are not allowed ('plugin_type', 'repo' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['transformers']['items']:
    {'$ref': '#/$defs/plugins/plugin_generic',
     'description': 'https://docs.meltano.com/concepts/plugins#transforms',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['transformers'][0]:
    {'commands': {'build': {'args': 'build',
                            'description': 'Will run your models, tests, '
                                           'snapshots and seeds in DAG '
                                           'order.'},
                  'clean': {'args': 'clean',
                            'description': 'Delete all folders in the '
                                           'clean-targets list (usually '
                                           'the dbt_modules and target '
                                           'directories.)'},
                  'compile': {'args': 'compile',
                              'description': 'Generates executable SQL '
                                             'from source model, test, and '
                                             'analysis files. Compiled SQL '
                                             'files are written to the '
                                             'target/ directory.'},
                  'debug': {'args': 'debug',
                            'description': 'Debug your DBT project and '
                                           'warehouse connection.'},
                  'deps': {'args': 'deps',
                           'description': 'Pull the most recent version of '
                                          'the dependencies listed in '
                                          'packages.yml'},
                  'docs-generate': {'args': 'docs generate',
                                    'description': 'Generate documentation '
                                                   'for your project.'},
                  'docs-serve': {'args': 'docs serve',
                                 'description': 'Serve documentation for '
                                                'your project. Make sure '
                                                'you ran `docs-generate` '
                                                'first.'},
                  'freshness': {'args': 'source freshness',
                                'description': 'Check the freshness of '
                                               'your source data.'},
                  'run': {'args': 'run',
                          'description': 'Compile SQL and execute against '
                                         'the current target database.'},
                  'seed': {'args': 'seed',
                           'description': 'Load data from csv files into '
                                          'your data warehouse.'},
                  'snapshot': {'args': 'snapshot',
                               'description': 'Execute snapshots defined '
                                              'in your project.'},
                  'test': {'args': 'test',
                           'description': 'Runs tests on data in deployed '
                                          'models.'}},
     'docs': 'https://docs.meltano.com/guide/transformation',
     'env': {'DBT_POSTGRES_PROFILES_DIR': '/home/will/meltano/projects/sandbox/transform/profiles/postgres',
             'DBT_POSTGRES_PROJECT_DIR': '/home/will/meltano/projects/sandbox/transform',
             'DBT_PROFILES_DIR': '/home/will/meltano/projects/sandbox/transform/profiles/postgres',
             'MELTANO_TRANSFORM_PROFILES_DIR': '/home/will/meltano/projects/sandbox/transform/profiles/postgres',
             'MELTANO_TRANSFORM_PROJECT_DIR': '/home/will/meltano/projects/sandbox/transform'},
     'executable': 'dbt',
     'label': 'dbt PostgreSQL',
     'name': 'dbt-postgres',
     'namespace': 'dbt_postgres',
     'pip_url': 'dbt-core~=1.1.0 dbt-postgres~=1.1.0',
     'plugin_type': 'transformers',
     'repo': 'https://github.com/dbt-labs/dbt-core',
     'requires': {'files': [{'name': 'files-dbt-postgres',
                             'variant': 'meltano'}]},
     'settings': [{'label': 'Projects Directory',
                   'name': 'project_dir',
                   'value': '$MELTANO_PROJECT_ROOT/transform'},
                  {'env': 'DBT_PROFILES_DIR',
                   'label': 'Profiles Directory',
                   'name': 'profiles_dir',
                   'value': '$MELTANO_PROJECT_ROOT/transform/profiles/postgres'},
                  {'description': 'The postgres host to connect to.\n',
                   'kind': 'string',
                   'label': 'Host',
                   'name': 'host'},
                  {'description': 'The user to connect as.\n',
                   'kind': 'string',
                   'label': 'User',
                   'name': 'user'},
                  {'description': 'The password to connect with.\n',
                   'kind': 'password',
                   'label': 'Password',
                   'name': 'password'},
                  {'description': 'The port to connect to.\n',
                   'kind': 'integer',
                   'label': 'Port',
                   'name': 'port'},
                  {'aliases': ['database'],
                   'description': 'The db to connect to.\n',
                   'kind': 'string',
                   'label': 'Database',
                   'name': 'dbname'},
                  {'description': 'The schema to use.\n',
                   'kind': 'string',
                   'label': 'Schema',
                   'name': 'schema'},
                  {'description': 'Seconds between TCP keepalive '
                                  'packets.\n',
                   'kind': 'integer',
                   'label': 'Keep Alives Idle',
                   'name': 'keepalives_idle'},
                  {'description': 'Overrides the default search path.\n',
                   'kind': 'string',
                   'label': 'Search Path',
                   'name': 'search_path'},
                  {'description': 'Role for dbt to assume when executing '
                                  'queries.\n',
                   'kind': 'string',
                   'label': 'Role',
                   'name': 'role'},
                  {'description': 'SSL Mode used to connect to the '
                                  'database.\n',
                   'kind': 'array',
                   'label': 'SSL Mode',
                   'name': 'sslmode'}],
     'variant': 'dbt-labs'}
Unevaluated properties are not allowed ('plugin_type', 'repo', 'update' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['files']['items']:
    {'$ref': '#/$defs/plugins/plugin_generic',
     'description': 'https://docs.meltano.com/concepts/plugins#file-bundles',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['files'][0]:
    {'docs': 'https://hub.meltano.com/files/files-airflow',
     'env': {'FILES_AIRFLOW__UPDATE': '{"orchestrate/dags/meltano.py": '
                                      'true}',
             'FILES_AIRFLOW__UPDATE_ORCHESTRATE_DAGS_MELTANO_PY': 'true',
             'MELTANO_FI__UPDATE': '{"orchestrate/dags/meltano.py": true}',
             'MELTANO_FI__UPDATE_ORCHESTRATE_DAGS_MELTANO_PY': 'true'},
     'label': 'Airflow',
     'name': 'files-airflow',
     'namespace': 'files_airflow',
     'pip_url': 'git+https://github.com/meltano/files-airflow.git',
     'plugin_type': 'files',
     'repo': 'https://github.com/meltano/files-airflow',
     'update': {'orchestrate/dags/meltano.py': True},
     'variant': 'meltano'}
Unevaluated properties are not allowed ('plugin_type', 'repo' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['utilities']['items']:
    {'$ref': '#/$defs/plugins/plugin_generic',
     'description': 'https://docs.meltano.com/concepts/plugins#utilities',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['utilities'][0]:
    {'commands': {'fix': {'args': 'fix',
                          'description': 'Automatically fix SQL files '
                                         'according to your defined '
                                         'rules.'},
                  'fix-force': {'args': 'fix --force',
                                'description': 'Automatically fix SQL '
                                               'files according to your '
                                               'defined rules (does not '
                                               'ask for confirmation).'},
                  'lint': {'args': 'lint',
                           'description': 'Lint SQL in transform models.'}},
     'docs': 'https://hub.meltano.com/utilities/sqlfluff',
     'env': {},
     'label': 'SQLFluff',
     'name': 'sqlfluff',
     'namespace': 'sqlfluff',
     'pip_url': 'sqlfluff[dbt]',
     'plugin_type': 'utilities',
     'repo': 'https://github.com/sqlfluff/sqlfluff',
     'variant': 'sqlfluff'}
Unevaluated properties are not allowed ('plugin_type' was unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['utilities']['items']:
    {'$ref': '#/$defs/plugins/plugin_generic',
     'description': 'https://docs.meltano.com/concepts/plugins#utilities',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['utilities'][1]:
    {'commands': {'describe': {'args': 'describe', 'executable': 'cron'}},
     'env': {},
     'executable': 'cron',
     'label': 'cron',
     'name': 'cron',
     'namespace': 'cron',
     'pip_url': '-e $CRON_PIP_LOCATION',
     'plugin_type': 'utilities'}
Unevaluated properties are not allowed ('plugin_type', 'repo' were unexpected)

Failed validating 'unevaluatedProperties' in schema['properties']['plugins']['properties']['transforms']['items']:
    {'allOf': [{'$ref': '#/$defs/plugins/plugin_generic'},
               {'properties': {'package_name': {'description': 'The name '
                                                               'of the dbt '
                                                               "package's "
                                                               'internal '
                                                               'dbt '
                                                               'project: '
                                                               'the value '
                                                               'of name in '
                                                               'dbt_project.yml.',
                                                'type': 'string'},
                               'vars': {'description': 'An object '
                                                       'containing dbt '
                                                       'model variables',
                                        'type': 'object'}}}],
     'description': 'https://docs.meltano.com/concepts/plugins#transforms',
     'type': 'object',
     'unevaluatedProperties': False}

On instance['plugins']['transforms'][0]:
    {'docs': 'https://hub.meltano.com/transforms/tap-google-analytics--meltano',
     'env': {'MELTANO_TRANSFORM__PACKAGE_NAME': 'tap_google_analytics',
             'MELTANO_TRANSFORM__VARS': '{"schema": "{{ '
                                        "env_var('DBT_SOURCE_SCHEMA','tap_google_analytics') "
                                        '}}"}',
             'MELTANO_TRANSFORM__VARS_SCHEMA': '{{ '
                                               "env_var('DBT_SOURCE_SCHEMA','tap_google_analytics') "
                                               '}}',
             'TAP_GOOGLE_ANALYTICS__PACKAGE_NAME': 'tap_google_analytics',
             'TAP_GOOGLE_ANALYTICS__VARS': '{"schema": "{{ '
                                           "env_var('DBT_SOURCE_SCHEMA','tap_google_analytics') "
                                           '}}"}',
             'TAP_GOOGLE_ANALYTICS__VARS_SCHEMA': '{{ '
                                                  "env_var('DBT_SOURCE_SCHEMA','tap_google_analytics') "
                                                  '}}'},
     'label': 'tap-google-analytics',
     'name': 'tap-google-analytics',
     'namespace': 'tap_google_analytics',
     'pip_url': 'https://gitlab.com/meltano/dbt-tap-google-analytics.git@config-version-2',
     'plugin_type': 'transforms',
     'repo': 'https://gitlab.com/meltano/dbt-tap-google-analytics',
     'variant': 'meltano',
     'vars': {'schema': '{{ '
                        "env_var('DBT_SOURCE_SCHEMA','tap_google_analytics') "
                        '}}'}}
'5 * */2 * 5,6,0' does not match '^((@(hourly|daily|weekly|monthly|yearly|once))|((((\\d+,)+\\d+|(\\d+(\\/|-)\\d+)|\\d+|\\*) ?){5,6}))$'

Failed validating 'pattern' in schema['properties']['schedules']['items']['properties']['interval']:
    {'description': 'A UNIX cron expression to represent the frequency the '
                    'scheduled job should execute',
     'examples': ['@hourly', '@daily', '@weekly', '0 0 * * *'],
     'pattern': '^((@(hourly|daily|weekly|monthly|yearly|once))|((((\\d+,)+\\d+|(\\d+(\\/|-)\\d+)|\\d+|\\*) '
                '?){5,6}))$',
     'type': 'string'}

On instance['schedules'][1]['interval']:
    '5 * */2 * 5,6,0'
Compiled /home/will/meltano/projects/sandbox/.meltano/manifests/meltano-manifest.prod.json
```

I knew calling `check-jsonschema` from the CLI produced much cleaner output, and I didn't want to reimplement their formatting code, so I opted to just call their CLI tool. Invoking a subprocess is a bit costly, and inelegant, but I think it was the most effective way to get good validation error messages.

Reply:
That makes sense, thanks for the explanation 🙂